api-authentication/server/index.js

```diff
@@ -0,0 +1,109 @@
+const errorMiddleware = require('./error-middleware');
+
+const db = new pg.Pool({ // eslint-disable-line
+  connectionString: process.env.DATABASE_URL
+});
+
+const app = express();
+
+const jsonMiddleware = express.json();
+
+app.use(jsonMiddleware);
+
+app.post('/api/auth/sign-up', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(400, 'username and password are required fields');
+  }
+  argon2
+    .hash(password)
+    .then(hashedPassword => {
+      const sql = `
+        insert into "users" ("username", "hashedPassword")
+        values ($1, $2)
+        returning "userId", "username", "createdAt"
+      `;
+      const params = [username, hashedPassword];
+      return db.query(sql, params);
+    })
+    .then(result => {
+      const [user] = result.rows;
+      res.status(201).json(user);
+    })
+    .catch(err => next(err));
+});
+
+app.post('/api/auth/sign-in', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(401, 'invalid login');
+  }
+
+  /* your code starts here */
+
+  /**
+   * Query the database to find the "userId" and "hashedPassword" for the "username".
+   * Then, 😉
+   *    If no user is found,
+   *      throw a 401: 'invalid login' error.
+   *    If a user is found,
+   *      confirm that the password included in the request body matches the "hashedPassword" with `argon2.verify()`
+   *      Then, 😉
+   *        If the password does not match,
+   *          throw a 401: 'invalid login' error.
+   *        If the password does match,
+   *          Create a payload object containing the user's "userId" and "username".
+   *          Create a new signed token with `jwt.sign()`, using the payload and your TOKEN_SECRET
+   *          Send the client a 200 response containing the payload and the token.
+   *      Catch any error.
+   * Catch any error.
+   */
+
+  const sql = `select "userId", "hashedPassword" from "users"
+                where "username" = $1`;
+
+  const param = [username];
+
+  db.query(sql, param)
+    .then(result => {
+      if (result.rows.length === 0) {
+        throw new ClientError(401, 'Invalid login');
+      }
+
+      return argon2.verify(result.rows[0].hashedPassword, password)
+        .then(isMatching => {
+          if (isMatching) {
+            const payload = {
+              userId: result.rows[0].userId,
+              username: username
+            };
+
+            const token = jwt.sign(payload, process.env.TOKEN_SECRET);
+
+            const signedToken = {
+              token: token,
+              user: payload
+            };
+            res.status(200).send(signedToken);
+          } else {
+            throw new ClientError(401, 'Invalid login');
+          }
+        })
+        .catch(err => console.error('Error: ', err));
+    })
+    .catch(err => console.error('Error: ', err));
```

Comment:
Instead of logging the error, pass it to `next()` since the `errorMiddleware` automatically logs errors and responds to the client.